### PR TITLE
WIP fix codegen

### PR DESCRIPTION
**Pull request by gouarin**
This PR is a fix and improvement of codegen.

* codegen
  * fix the recursion when you have mutliple CodeBlocks
  * support other Matrix types
  * remove datatype function and define each datatype inside codegen classes for more flexibility
  * add settings in Routine to manage the printer for each routine when you have multiple
  * add idx_vars in Routine that stores the indexes contained in For (they need particular treatment sometimes)
* utilities
    * autowrap: wrap several routines in one go and export the module
    * autowrap: add NumPy and Pure Cython wrapper

This short example shows the issue for the recursion of For

```
x, y = symbols('x,y', real=True) 
i = Idx('i') 
test = For(i, Range(10), [Assignment(x, y)]) 
routine = make_routine('test', test) 
code_gen = C89CodeGen() 
source = get_string(code_gen.dump_c, [routine]) 
print(routine) 
print(source)
```

The output is

```
Routine('test', [InputArgument(x), InputArgument(y)], [Result(For(i, Range(0, 10, 1), CodeBlock(Assignment(x, y))), result_6824502057046278987, result_6824502057046278987)], {i}, set())
#include "file.h"
#include <math.h>
double test(double x, double y) {
   double test_result;
   test_result = for (i = 0; i < 10; i += 1) {
      x = y;
   };
   return test_result;
}
```

The problem is that For expression has a CodeBlock (list of statements) which can contain important information for output parameters and return values.

So, the idea is to inspect output parameters and result values recursively and not to see a routine as a result but as a list of statements where you can have results, input, output, input/output.

Now, the main issue is to change the sympy expression where you have a result parameter like in this expression

```
For(i, Range(10), [x*y])
```

The idea would be to replace `x*y` by `Assignment(result.name, x*y)`. In that way, all the work done for printing in a target language should be the same. 

I don't know how to do this correctly because each method `subs`, `replace`, ... returns a copy and we need to change tuple element of CodeBlock for example.

If you have some ideas ...

### CODE REVIEW

**Comment by bjodah** [comment]
I guess this should have a test added for the new capability?
Fixing `sympy.utilities.codegen` is of course great since `sympy.codegen.ast` doesn't offer a polished
interface yet. I just pushed a commit to #13100 where a test case builds and calls into a C-function with in-/output arguments. But it still requires a bit of boilerplate since `.codegen.ast` is mostly building blocks.

**Comment by gouarin** [comment]
@bjodah the codegen utilities code will be remove after your PR will be merged ? Is that the idea ?

**Comment by bjodah** [comment]
No, not in the foreseeable future, first `sympy.codegen` would need to reach feature completeness, and then there would be a long deprecation cycle, and even then it might not even be removed if the APIs differ so much that users cannot be expected to easily transition.

**Comment by gouarin** [comment]
Ok. For now, I have something which works for pyLBM (http://pylbm.readthedocs.io/en/develop/) but with an old version of sympy and I would like all the stuff I added to be in sympy to not have multiple versions and restart from the beginning.

I have to improve the packaging of pyLBM and for that I will need to have a dependency of a release of sympy with a working codegen.

**Comment by gouarin** [comment]
I also have a lot of numerical examples which can improve the idea of using sympy to build numerical codes and be able to have a robust codegen for multiple languages.

**Comment by asmeurer** [comment]
I have no comments on the changes here, but as a general comment, it would be useful for the codegen ast nodes to be able to specify how they should be interpreted as the rhs of an assignment. I guess this has to be done at the printer level, since different languages do different things. For a CodeBlock you probably want to assign the last line. For `For` you may want something similar (or maybe it should be an error, I don't know).

Forgive me if we already discussed this before, I can't remember.

**Comment by gouarin** [comment]
I'm not sure I understand what you mean. The assignment is not necessary after the last line of a CodeBlock. For example, you can have

`For(i, Range(10), [x*y, For(j, ... do something on x)])`

where `x` could be a local variable. I had a lot of problem with this `result` term when I hacked for the first time codegen. So, I decided to remove it but I know it is not the solution.

The best solution for me is to ask to the user to declare the result variables as `Result` and then, doing the assignment like any other variables.
@asmeurer @bjodah tell me if you are agree with that.

The other problem that I encountered was the index variables in a For loop: they must be local variables which is not the case for now. I can also fix that.

For me, ast should be like an intermediate representation (like LLVM does) where it is easy to add any printing language class on it.

**Comment by gouarin** [comment]
@asmeurer @bjodah could you give me your opinion about let user to define `Result` and not to try to predict it  as mentioned in my last message ?

**Comment by bjodah** [comment]
@gouarin you are right that we have quite a bit of degree of freedom to assume things in codegen.
But your example:
```
x, y = symbols('x,y', real=True) 
i = Idx('i') 
test = For(i, Range(10), [Assignment(x, y)]) 
```
in this case I believe inferring x and y being arrays is taking those assumptions one step too far.
We have an `Indexed` class in `sympy.tensor.indexed` that maps quite well to C/Fortran-arrays.

**Comment by gouarin** [comment]
@bjodah my example was not a good one.

I will modify codegen with the assumption that a user must write if he want a `Result` and not to give just an expression without assignment.

Then, I will let you tell me what do you think about this modifications.

**Comment by gouarin** [comment]
@asmeurer @bjodah I updated this PR to have codegen more readable and more flexible.

Now, I will add tests for loop code using `For` but it should work easily with what I have done.

Tell me if you are ok with theses changes.

File: sympy/utilities/tests/test_codegen_julia.py [hunk 1]
```diff
@@ -488,7 +488,7 @@ def test_jl_loops():
                       header=False, empty=False)
     source = result[1]
     expected = (
-        'function mat_vec_mult(y, A, m, n, x)\n'
+        'function mat_vec_mult(A, m, n, x, y)\n'
```
Comment (bjodah) [inline sympy/utilities/tests/test_codegen_julia.py]: This looks like it could be a breaking change, no?

Comment (gouarin) [reply sympy/utilities/tests/test_codegen_julia.py]: it's only in Julia where the outputs are at the beginning. I don't know why this choice... It doesn't make sens to me.

But I am a newbie in Julia.

**Comment by Abdullahjavednesar** [comment]
ping @bjodah.

**Comment by moorepants** [comment]
What is the status on this PR? It is labeled as WIP, so maybe still there.

**Comment by gouarin** [comment]
I didn't have time to finish this PR but I need it for my package. There is
a lot of work to be done for the autowrap module.

I plan to finish it before the end of this year.



Le dim. 7 oct. 2018 à 13:48, Jason K. Moore <notifications@github.com> a
écrit :

**Comment by moorepants** [comment]
@gouarin Thanks for the update. We will leave it open.

**Comment by gouarin** [comment]
@bjodah @asmeurer @moorepants 

I think the improvement of codegen is done for me. Before I go further, I have some questions

- codegen and autowrap are now difficult to read and I would like to split them, what do you think ?
- I don't like these lines

https://github.com/sympy/sympy/blob/master/sympy/utilities/autowrap.py#L504-L507
https://github.com/sympy/sympy/blob/master/sympy/utilities/autowrap.py#L611-L615
https://github.com/sympy/sympy/blob/master/sympy/utilities/codegen.py#L1956-L1959

because we loose the flexibility that we could have with codegen to add other wrappers. I have no idea at this time to have a better format.
- For codegen and autowrap testing, I will try to implement in SymPy various numerical examples found here
 https://github.com/serge-sans-paille/numpy-benchmarks

and I will probably add Pythran and Numba wrappers since I started to put NumPy wrapper. Do I add these benchmarks in SymPy repo or do you prefer that I create a new one ?

I made a lot of changes so tell me also if it's ok for you. I think we can do a lot of things now with codegen.

File: sympy/printing/tests/test_ccode.py [hunk 2]
```diff
@@ -571,7 +571,7 @@ def test_ccode_Assignment():
 
 def test_ccode_For():
     f = For(x, Range(0, 10, 2), [aug_assign(y, '*', x)])
-    assert ccode(f) == ("for (x = 0; x < 10; x += 2) {\n"
+    assert ccode(f) == ("for (int x = 0; x < 10; x += 2) {\n"
```
Comment (bjodah) [inline sympy/printing/tests/test_ccode.py]: This is a breaking change (in general we shouldn't change existing tests, they are there not only to ensure correctness, but also to guarantee a stable API for end users). Besides, I don't think that generating code for `For` should not imply declaration of the loop variable in C (it's not even valid C89, which is still what e.g. msvc targets and what most embedded systems use). You'll have to generate the declaration and initialization outside the printer...

File: sympy/utilities/autowrap.py [hunk 3]
```diff
@@ -527,8 +609,8 @@ def _validate_backend_language(backend, language):
 
 @cacheit
 @doctest_depends_on(exe=('f2py', 'gfortran'), modules=('numpy',))
-def autowrap(expr, language=None, backend='f2py', tempdir=None, args=None,
-             flags=None, verbose=False, helpers=None, code_gen=None, **kwargs):
+def autowrap(expr_or_routines, language=None, backend='f2py', tempdir=None, args=None,
+             flags=None, verbose=False, helpers=None, code_gen=None, user_local_vars=None, settings={}, **kwargs):
```
Comment (bjodah) [inline sympy/utilities/autowrap.py]: we should not use [mutable default values](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) keyword arguments (they are essentially global variables in python, unfortunately). You can set de default to `None` and then upon first access use the idiom `settings or {}`.

File: sympy/utilities/tests/test_autowrap.py [hunk 4]
```diff
@@ -91,7 +91,7 @@ def test_cython_wrapper_inoutarg():
 def test_cython_wrapper_compile_flags():
     from sympy import Equality
     x, y, z = symbols('x,y,z')
-    routine = make_routine("test", Equality(z, x + y))
+    routine = [make_routine("test", Equality(z, x + y))]
```
Comment (bjodah) [inline sympy/utilities/tests/test_autowrap.py]: my earlier comment on not changing any tests applies here as well.

**Review by bjodah — COMMENTED** [review]
I left some comments on things that stood out to me. In general I think it's fine to refactor these modules heavily given that we don't break users code, i.e. we shouldn't modify existing tests (I hope that the test suite has sufficient coverage but that's another issue).

File: sympy/utilities/tests/test_autowrap.py (hunk 4 above)
Comment (gouarin) [reply sympy/utilities/tests/test_autowrap.py]: The problem is that `_prepare_files` takes now a list of routines so I can move the construction of this list when `_prepare_files` is called.

**Comment by gouarin** [comment]
I know that this PR begins to be huge and probably difficult to review. All I have done here, I use it in a real application https://github.com/pylbm/pylbm where the numerical code is generated by codegen.

Tell me how we can make progress to add these features in SymPy. We can merge (when you will be agree) and open an other PR that will split the codegen files.

I can also illustrate what I have done with some examples.

**Comment by bjodah** [comment]
You're not really just splitting files here (which is what I thought you were asking about on gitter), you're moving whole submodules. That might be ok, but I'd rather review such changes in a separate PR. I don't know what @asmeurer think though.

File: sympy/printing/tests/test_ccode.py (hunk 2 above)
Comment (bjodah) [reply sympy/printing/tests/test_ccode.py]: this is still a breaking change...

**Comment by asmeurer** [comment]
I agree that code rearrangements should go in a separate PR. That will make it much easier to review. Otherwise it is hard to find the new changes that sit alongside the rearrangements. 

I haven't looked at what the suggested rearrangements are, but we should try to keep backwards compatibility for the imports, especially for the things that aren't in `__init__.py`.

File: sympy/utilities/tests/test_autowrap.py (hunk 4 above)
Comment (asmeurer) [reply sympy/utilities/tests/test_autowrap.py]: Does this break public API?

Comment (bjodah) [reply sympy/utilities/tests/test_autowrap.py]: Oh, my mistake, it was only used in `_prepare_files` - so I guess no.

**Comment by gouarin** [comment]
@asmeurer @bjodah My PR is almost done but I have a last issue and I don't know how to solve it.

Suppose you have

```
A = MatrixSymbol('A', 8, 7)
B = MatrixSymbol('B', 2, 2)
name_expr = ("test", [Equality(B, A[6:, 1::3])])
```

For now, the generated code in SymPy is something like (for example for Julia)

```
function test(A)
    B = A[7:end,2:3:end]
    return B
end
```

But imagine you do that a lot of time, do you really want to allocate B at each call? You will prefer

```
function test(B, A)
    B = A[7:end,2:3:end]
end
```
It's the same for all other languages. The problem is I didn't find an easy way to accept both without mention somewhere that I don't want return values.

Do you have ideas to manage this two cases ?

**Comment by asmeurer** [comment]
I have very little experience with autowrap (what I'm assuming you are referring to), but doesn't the `args` argument let you specify arguments to the function?

**Comment by gouarin** [comment]
What do you mean by `args` ? 

When you extract all the elements of the expressions, you finally have `InputArguments`, `OutputArguments`, and `InOutArguments`. Depending on the language you use, you write the code by setting these parameters as input or as output of the function. But unfortunately, you can't make the decision where you want to have them.

File: sympy/codegen/ast.py [hunk 5]
```diff
@@ -785,7 +787,7 @@ def cse(self, symbols=None, optimizations=None, postprocess=None,
         return self.topological_sort(new_assignments + new_block)
 
 
-class For(Token):
+class For(Token, WithBody):
```
[the remark this answers is not in the document]
Comment (gouarin) [reply sympy/codegen/ast.py]: This is exactly for users who want to add their own `Token` which have a `Body`. Then codegen must be recursive to analyse all the expressions.

File: sympy/printing/pycode.py [hunk 6]
```diff
@@ -610,6 +612,109 @@ def _print_CodegenArrayPermuteDims(self, expr):
...
+                        upper += a.upper
+                    else:
+                        lower += a
+                        upper += a
+                elem.append("%s:%s"%(lower, upper))
+            else:
+                elem.append(self._print(expr.indices[i]))
+        return "%s[%s]" % (self._print(expr.base.label), ', '.join(elem))
+
+    def _print_Idx(self, expr):
```
Comment (bjodah) [inline sympy/printing/pycode.py]: I think `PythonCodePrinter` should behave the same way as `CCodePrinter` whenever possible.

File: sympy/printing/pycode.py [hunk 7]
```diff
@@ -610,6 +612,109 @@ def _print_CodegenArrayPermuteDims(self, expr):
...
+        return "%s:%s"%(expr.lower, expr.upper)
+
+    def _print_For(self, expr):
```
Comment (bjodah) [inline sympy/printing/pycode.py]: same here, `_print_For` should print a functional for loop should it not? (I don't see `for ... in ...:` being printed?)

Comment (gouarin) [reply sympy/printing/pycode.py]: It's not `PythonCodePrinter`but `NumPyCodePrinter`you read. And for the `For` loop you want the vectorized form of NumPy. This is why you don't have `for ... in ...`

File: sympy/codegen/ast.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bjodah) [reply sympy/codegen/ast.py]: I see, should we come up with a plausible use case and add it to the tests and the documentation in that case?

File: sympy/printing/pycode.py (hunk 7 above)
Comment (bjodah) [reply sympy/printing/pycode.py]: I see, we should probably add independet tests to `test_pycode.py`.

File: sympy/codegen/ast.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gouarin) [reply sympy/codegen/ast.py]: I begin to add various tests with autowrap which call the generated code and test if the result is as expected. But the main problem is, as I mentioned, where we set the output: in the parameters or as a result.

[the remark this answers is not in the document]
Comment (gouarin) [reply sympy/codegen/ast.py]: In my application, I use it as a parameter. Almost never as a result.

File: sympy/printing/pycode.py (hunk 7 above)
Comment (gouarin) [reply sympy/printing/pycode.py]: This is planned. Sorry to not push yet the test cases. But I want them to be clean.
I also try to keep the previous test unchanged as you wanted. I just have one to fix.

**Comment by czgdp1807** [comment]
@gouarin Thanks for the contributions. Would you still like to continue your work on this PR?

**Comment by gouarin** [comment]
@czgdp1807 yes, I want to finish this PR. I let the previous SymPy tests unchanged and add the possibility to test codegen output with compilation, execution, and output verification.

I just have one thing to add. The possibility to put the output of a function into the parameters or as a return. Most of my usage is into the parameters. The idea is to not allocate an array and return it at each function call (it's just an example). But the default behavior of codegen is to put the output as a return.

And, I am really sorry for the delay !!

**Comment by czgdp1807** [comment]
@certik @bjodah @asmeurer Any views on the above comment?

**Comment by czgdp1807** [comment]
@gouarin Please resolve the conflicts. Please do not close the PR.

**Comment by gouarin** [comment]
@czgdp1807 I will resolve the conflicts next week and I will do my best to finish this PR before the end of June.

**Comment by gouarin** [comment]
I finally decided to take a new start and split this PR into smallest to facilitate the review.

I plan to make the following PR

- fix codegen for nested blocks (https://github.com/sympy/sympy/pull/20877)
- add the possibility to have several routines constructed by autowrap
- add NumPy and pure Cython codegen
- add tests into external directory to test both NumPy and pure Cython generator

**Pull request closed without merging**